Python/optimizer.c

```diff
@@ -1258,6 +1287,48 @@ int effective_trace_length(_PyUOpInstruction *buffer, int length)
 }
 #endif
 
+static int
+stack_allocate(_PyUOpInstruction *buffer, int length)
+{
+    for (int i = length-1; i >= 0; i--) {
+        buffer[i*2].format = UOP_FORMAT_TARGET;
+        buffer[i*2].oparg = 0;
+        buffer[i*2].target = 0;
+        buffer[i*2+1] = buffer[i];
+    }
+    int depth = 0;
+    for (int i = 0; i < length; i++) {
+        _PyUOpInstruction *spill_or_reload = &buffer[i*2];
+        int uop = buffer[i*2+1].opcode;
+        if (uop == _NOP) {
+            // leave _NOPs to be cleaned up later
+            spill_or_reload->opcode = _NOP;
+            continue;
+        }
+        if (_PyUop_Caching[uop].opcodes[depth] > 0) {
+            spill_or_reload->opcode = _NOP;
+        }
+        else {
+            int min_depth = _PyUop_Caching[uop].min_input;
+            int max_depth = _PyUop_Caching[uop].max_input;
+            if (depth < min_depth) {
+                spill_or_reload->opcode = _PyUop_SpillsAndReloads[depth][min_depth];
+                depth = min_depth;
+            }
+            else {
+                assert(depth != max_depth);
+                spill_or_reload->opcode = _PyUop_SpillsAndReloads[depth][max_depth];
+                depth = max_depth;
+            }
+        }
+        assert(_PyUop_Caching[uop].opcodes[depth] != 0);
+        assert(spill_or_reload->opcode != 0);
+        buffer[i*2+1].opcode = _PyUop_Caching[uop].opcodes[depth];
```

Comment:
It would be good to add custom cases to better handle COPY and SWAP later via replication. But we can do that in another PR.